configuration/azure/appconfig/appconfig.go

```diff
@@ -0,0 +1,190 @@
+)
+
+// ConfigurationStore is a Azure App Configuration store.
+type ConfigurationStore struct {
+	client   *azappconfig.Client
+	json     jsoniter.API
+	metadata metadata
+
+	logger logger.Logger
+}
+
+// NewAzureAppConfigurationStore returns a new Azure App Configuration store.
+func NewAzureAppConfigurationStore(logger logger.Logger) configuration.Store {
+	s := &ConfigurationStore{
+		json:   jsoniter.ConfigFastest,
+		logger: logger,
+	}
+
+	return s
+}
+
+// Init does metadata and connection parsing.
+func (r *ConfigurationStore) Init(metadata configuration.Metadata) error {
+	m, err := parseMetadata(metadata)
+	if err != nil {
+		return err
+	}
+	r.metadata = m
+
+	coreClientOpts := azcore.ClientOptions{
+		Telemetry: policy.TelemetryOptions{
+			ApplicationID: "dapr-" + logger.DaprVersion,
+		},
+	}
+
+	options := azappconfig.ClientOptions{
+		ClientOptions: coreClientOpts,
+	}
+
+	if r.metadata.connectionString != "" {
+		r.client, err = azappconfig.NewClientFromConnectionString(connectionString, &options)
+		if err != nil {
+			return err
+		}
+	} else {
+		var settings azauth.EnvironmentSettings
+		settings, err = azauth.NewEnvironmentSettings("appconfig", metadata.Properties)
+		if err != nil {
+			return err
+		}
+
+		var cred azcore.TokenCredential
+		cred, err = settings.GetTokenCredential()
+		if err != nil {
+			return err
+		}
+
+		r.client, err = azappconfig.NewClient(r.metadata.host, cred, &options)
+		if err != nil {
+			return err
+		}
+	}
+
+	return err
+}
+
+func parseMetadata(meta configuration.Metadata) (metadata, error) {
+	m := metadata{}
+
+	if val, ok := meta.Properties[host]; ok && val != "" {
+		m.host = val
+	} else {
+		return m, errors.New("redis store error: missing host address")
+	}
+
+	if val, ok := meta.Properties[connectionString]; ok && val != "" {
+		m.connectionString = val
+	}
+
+	m.maxRetries = defaultMaxRetries
+	if val, ok := meta.Properties[maxRetries]; ok && val != "" {
+		parsedVal, err := strconv.Atoi(val)
+		if err != nil {
+			return m, fmt.Errorf("redis store error: can't parse maxRetries field: %s", err)
+		}
+		m.maxRetries = parsedVal
+	}
+
+	m.maxRetryDelay = defaultMaxRetryDelay
+	if val, ok := meta.Properties[maxRetryDelay]; ok && val != "" {
+		parsedVal, err := strconv.Atoi(val)
+		if err != nil {
+			return m, fmt.Errorf("redis store error: can't parse maxRetryDelay field: %s", err)
+		}
+		m.maxRetryDelay = time.Duration(parsedVal)
+	}
+
+	m.retryDelay = defaultRetryDelay
+	if val, ok := meta.Properties[retryDelay]; ok && val != "" {
+		parsedVal, err := strconv.Atoi(val)
+		if err != nil {
+			return m, fmt.Errorf("redis store error: can't parse retryDelay field: %s", err)
+		}
+		m.retryDelay = time.Duration(parsedVal)
+	}
+
+	return m, nil
+}
+
+func (r *ConfigurationStore) Get(ctx context.Context, req *configuration.GetRequest) (*configuration.GetResponse, error) {
+	keys := req.Keys
+	items := make([]*configuration.Item, 0, len(keys))
+	for _, key := range keys {
```

Comment:
The only problem I see here is that we're making requests to the service one at a time.

This method is primarily used to retrieve the initial config state at startup.

Assume that 2 apps are starting up and are requesting the configuration. In the meanwhile, someone is changing the configuration.

Because there's no locking, it's possible that the two apps could get a different configuration as they loop through all keys.

Is there no way to get all keys at once, atomically?

Reply:
@ItalyPaleAle  there seems to be no way to get everything at once.

We could use the [NewListRevisionsPager](https://github.com/Azure/azure-sdk-for-go/tree/sdk/appconfig/azappconfig/v0.3.2/sdk/appconfig/azappconfig/#list-configuration-setting-revisions) method together with a [filter](https://docs.microsoft.com/en-us/azure/azure-app-configuration/rest-api-key-value#filtering) by key.

Given the 5 keys per filter limitations, code would end up making 1 call for each block of 5 or less keys present in the request.

Thoughts?

Reply:
Uhm, I don't think that would solve the issue.

I guess this is what it is. Maybe just make sure this behavior is documented?

Reply:
Referenced this review here: https://github.com/dapr/docs/issues/2660